ilastik/workflows/carving/opPreprocessing.py

```diff
@@ -182,40 +182,36 @@ def setupOutputs(self):
 
     def execute(self, slot, subindex, roi, result):
         assert roi.stop - roi.start == self.Output.meta.shape, "Blockwise Watershed must be run on the entire volume."
-        input_image = self.Input(roi.start, roi.stop).wait()
-        volume_feat = input_image[0, ..., 0]
-        result_view = result[0, ..., 0]
-        # handle the special case of the Request threadpool not having any workers
-        max_workers = max(1, Request.global_thread_pool.num_workers)
-        with Timer() as watershedTimer:
-            # 3d watersheds
-            if self.Input.meta.getTaggedShape()["z"] > 1:
-                logger.info("Run block-wise watershed in 3d")
-                if self.DoAgglo.value:
-                    result_view[...], max_id = watershed_and_agglomerate(
-                        volume_feat,
-                        max_workers=max_workers,
+
+        if self.Input.meta.getTaggedShape()["z"] > 1:
+            idx = numpy.s_[0, ..., 0]
+        else:
+            idx = numpy.s_[0, ..., 0, 0]
+
+        # Don't adjust roi to idx because idx should only drop singleton dims.
+        input_ = self.Input(roi.start, roi.stop).wait()[idx]
+        assert input_.ndim in 2, 3
+
+        with Timer() as timer:
+            logger.info("Run block-wise watershed in %sd", input_.ndim)
+
+            if self.DoAgglo.value:
+                result[idx], max_id = watershed_and_agglomerate(
+                        input_,
+                        max_workers=max(1, Request.global_thread_pool.num_workers),
                         size_regularizer=self.SizeRegularizer.value,
                         reduce_to=self.ReduceTo.value,
                     )
-                else:
-                    result_view[...], max_id = vigra.analysis.watershedsNew(volume_feat)
-            # 2d watersheds
+            elif input_.ndim == 3:
+                result[idx], max_id = vigra.analysis.watersheds(input_)
+            elif input_.ndim == 2:
+                result[idx], max_id = vigra.analysis.watershedsNew(input_)
```

Comment:
I don't think we need the two branches here, because we can just use `vigra.analysis.watershedsNew`.
AFAIK `watersheds` and `watershedsNew` give the same output, but `watershedsNew` should be faster.
My guess is that someone just didn't refactor the 3d code branch when `watershedsNew` was introduced...

Reply:
OK, great! I assumed that the new watershed performed poorly on 3D.

Reply:
No, I am pretty sure it performs better.
AFAIK the only reason `watersheds` is still in `vigra` is because it supports some options not available in `watershedsNew` (turbo watershed), that are not relevant for us here.